module/scmi_reset/src/mod_scmi_reset.c

```diff
@@ -0,0 +1,447 @@
+
+    return FWK_SUCCESS;
+}
+
+/*
+ * Given a service identifier, retrieve a pointer to its agent's
+ * \c mod_scmi_reset_agent structure within the agent table.
+ */
+static int get_agent_entry(fwk_id_t service_id,
+			   const struct mod_scmi_reset_agent **agent)
+{
+    int status = 0;
+    unsigned int agent_id = 0;
+
+    status = scmi_reset_ctx.scmi_api->get_agent_id(service_id, &agent_id);
+    if (status != FWK_SUCCESS)
+        return status;
+
+    if (agent_id >= scmi_reset_ctx.config->agent_count)
+        return FWK_E_PARAM;
+
+    *agent = &scmi_reset_ctx.agent_table[agent_id];
+
+    return FWK_SUCCESS;
+}
+
+static int get_reset_device(fwk_id_t service_id, unsigned int domain_id,
+                          const struct mod_scmi_reset_device **reset_device)
+{
+    int status = 0;
+    const struct mod_scmi_reset_agent *agent_entry = NULL;
+
+    status = get_agent_entry(service_id, &agent_entry);
+    if (status != FWK_SUCCESS)
+        return status;
+
+    if (domain_id >= agent_entry->device_count)
+        return FWK_E_RANGE;
+
+    *reset_device = &agent_entry->device_table[domain_id];
+
+    assert(fwk_module_is_valid_element_id((*reset_device)->element_id));
+    if (!fwk_module_is_valid_element_id((*reset_device)->element_id))
+	    return FWK_E_PANIC;
+
+    return FWK_SUCCESS;
+}
+
+/*
+ * Query the permissions for a service making a request to determine if
+ * the operation is permitted on the given clock device.
+ */
+static bool service_is_permitted(const struct mod_scmi_reset_device *device,
+                                enum mod_scmi_reset_permissions request)
+{
+    return device->permissions & request;
+}
+
+static int reset_attributes_handler(fwk_id_t service_id,
+                                    const uint32_t *payload)
+{
+    const struct mod_scmi_reset_device *reset_device = NULL;
+    struct scmi_reset_attributes_a2p params = { 0 };
+    struct scmi_reset_attributes_p2a outmsg = {
+        .status = SCMI_GENERIC_ERROR,
+    };
+    size_t outmsg_size = sizeof(outmsg.status);
+    int status = FWK_SUCCESS;
+
+    params = *(const struct scmi_reset_attributes_a2p *)payload;
+
+    status = get_reset_device(service_id, params.domain_id, &reset_device);
+    if (status != FWK_SUCCESS) {
+        outmsg.status = SCMI_NOT_FOUND;
+        goto exit;
+    }
+
+    if (!service_is_permitted(reset_device, MOD_SCMI_RESET_PERM_ATTRIBUTES)) {
+        outmsg.status = SCMI_DENIED;
+        goto exit;
+    }
+
+    /*
+     * TODO: get Async/Sync support info from module reset
+     * Currently: no support for async reset, no support for notification
+     */
+    outmsg.flags &= ~SCMI_RESET_DOMAIN_ATTR_ASYNC;
+    outmsg.flags &= ~SCMI_RESET_DOMAIN_ATTR_NOTIF;
+
+    /* TODO: Get reset latency from reset driver */
+    outmsg.latency = SCMI_RESET_DOMAIN_ATTR_UNK_LAT;
+
+    strncpy((char *)outmsg.name, fwk_module_get_name(reset_device->element_id),
+            sizeof(outmsg.name) - 1);
```

Comment:
Are we copying without a terminating NULL char?

Reply:
right, i will remove `- 1`.